### PR TITLE
[backport] Import `using_allocator` in `cupy.cuda`

### PR DESCRIPTION
**Pull request by emcastillo**
Backport of #2951

### CODE REVIEW

**Comment by emcastillo** [comment]
[automatic post] Jenkins, test this please.

**Comment by niboshi** [comment]
"no-compat" PRs cannot be backported.

We assumed this fix as not no-compat. (https://github.com/cupy/cupy/pull/2951#issuecomment-574123088)

File: cupy/cuda/memory.pyx [hunk 1]
```diff
@@ -581,8 +580,7 @@ cpdef get_allocator():
         return allocator
 
 
-@contextlib.contextmanager
-def using_allocator(allocator=None):
+def _using_allocator(allocator=None):
```
Comment (niboshi) [inline cupy/cuda/memory.pyx]: Why not just move this function to `cuda/__init__.py`?

**Comment by niboshi** [comment]
I'll make a PR for the above comment.

File: cupy/cuda/memory.pyx (hunk 1 above)
Comment (jakirkham) [reply cupy/cuda/memory.pyx]: I tried that, but it relies on implementation details only accessible in that module. Eventually I gave up and left it there adding a wrapper in `cuda/__init__.py`. After all my main interest was just getting it in the docs 😉

That said, if you are able to find a way to fix this, that would be great. 😄

File: cupy/cuda/__init__.py [hunk 2]
```diff
@@ -100,6 +100,27 @@ def is_available():
...
+        allocator (function): CuPy memory allocator. It must have the same
+            interface as the :func:`cupy.cuda.alloc` function, which takes the
+            buffer size as an argument and returns the device buffer of that
+            size. When ``None`` is specified, raw memory allocator will be
+            used (i.e., memory pool is disabled).
+
+    Note:
+        This wraps an internal version of this function to provide a
+        `contextmanager` as `contextmanger` decoration doesn't behave
+        well in Cython.
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply cupy/cuda/__init__.py]: That seems fair. Thanks for the feedback 🙂

File: cupy/cuda/memory.pyx (hunk 1 above)
Comment (niboshi) [reply cupy/cuda/memory.pyx]: Thank you for a comment!
I sent a fix as #3094.
I'm not 100% sure this is version is better, but please take a look.

**Pull request merged**